core/classes/Templates/AssetResolver.php

```diff
@@ -0,0 +1,147 @@
+<?php
+
+/**
+ * Template asset management class.
+ *
+ * @package NamelessMC\Templates
+ * @see TemplateBase
+ * @author Aberdeener
+ * @version 2.0.0-pr13
+ * @license MIT
+ */
+class AssetResolver extends AssetTree {
+
+    /**
+     * @var array<string> Array of assets currently resolved.
+     */
+    private array $_assets = [];
+
+    /**
+     * Add assets to be put in the response HTML.
+     *
+     * @param string|array<string> $assets The asset(s) to resolve and add. Must be a constant from the `AssetTree` class.
+     */
+    public function include($assets): void {
+        if (!is_array($assets)) {
+            $assets = [$assets];
+        }
+
+        foreach ($assets as $asset) {
+            $this->validateAsset($asset);
+
+            $this->_assets[$asset] = parent::ASSET_TREE[$asset];
+        }
+    }
+
+    /**
+     * Resolve all the assets as an array of CSS file and JS files to add to
+     * the template as `link` or `script` elements respectively.
+     *
+     * @return array<array<string>> The resolved assets ready to be added to the template.
+     */
+    public function compile(): array {
+        $css = [];
+        $js = [];
+
+        foreach ($this->_assets as $name => $asset) {
+            $this->gatherAsset($name, $asset, $css, $js);
+        }
+
+        return [$css, $js];
+    }
+
+    /**
+     * Generate URLs for the given asset, and add them to the CSS and JS URL arrays as needed.
+     * This will also resolve and gather dependencies for the asset if applicable.
+     *
+     * @param string $name The name of the asset.
+     * @param array $asset The asset to gather.
+     * @param array $css Array of CSS assets already resolved to add to.
+     * @param array $js Array of JS assets already resolved to add to.
+     */
+    private function gatherAsset(string $name, array $asset, array &$css, array &$js): void {
+        // Load the dependencies first so that they're the first to be added to the HTML
+        foreach ($asset['depends'] as $dependency) {
+            // Don't throw an exception if the dependency has already been resolved
+            // since it is a dependency, there is a high chance it has already been resolved anyway
+            if ($this->validateAsset($dependency, false)) {
+                $this->gatherAsset($dependency, parent::ASSET_TREE[$dependency], $css, $js);
+            }
+        }
+
+        // Unideal solution to get the correct light/dark Prism dependency
+        if ($name === parent::TINYMCE) {
```

Comment:
Might be worth creating an issue for this now so we can come back to it.

Perhaps an asset definition can optionally be presented as an array with a `dark` and `light` option